[cert_manager] Migrate from integrations-extras

davidor opened the pull request:
### What does this PR do?

Migrates the cert-manager integration from [integrations-extras](https://github.com/DataDog/integrations-extras).

I have not changed the code. All credit goes to @arapulido who created the integration and all the other contributors.

I just adapted a few things like the install instructions in the README, the maintainer email, etc.

This is my first time migrating an integration from the integrations-extras repo, so I have some questions:
- I have set 1.0.0 as the version number and reset the CHANGELOG. Is that OK or should we continue from the version in integrations-extras?
- I've changed the author to `Datadog` because I've seen that's what we use in the integrations of this repo. Is that OK or should we keep the original in this case?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached

sarah-witt commented on cert_manager/setup.py:
```diff
@@ -0,0 +1,64 @@
+# (C) Datadog, Inc. 2019-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from codecs import open  # To use a consistent encoding
```
Where is the `pyproject.toml`? Did you happen to migrate an older version of the integration? https://github.com/DataDog/integrations-extras/blob/master/cert_manager/pyproject.toml

sarah-witt replied:
Where is the `pyproject.toml`? Did you happen to migrate an older version of the integration? https://github.com/DataDog/integrations-extras/blob/master/cert_manager/pyproject.toml

urseberry reviewed (approved):
Prose looks good. I would copy over the CHANGELOG entries to preserve history, but I can see an argument for either option.

hithwen commented:
Can you add container integrations as codeowners for this?

davidor commented on cert_manager/pyproject.toml:
```diff
@@ -0,0 +1,62 @@
+    "Topic :: System :: Monitoring",
+]
+dependencies = [
+    "datadog-checks-base>=4.2.0",
```
@sarah-witt @hithwen I've noticed that the version required here is much older than in the rest of integrations. Should we keep this as it was or use a different version?

sarah-witt replied:
We should update the version to what other openmetrics checks use

sarah-witt reviewed (commented):
I took a look over the package format and left a few comments. The requirements.in file can also be removed now. Finally, this integration should have config models now that's in in core, but those can be added in a separate PR.

sarah-witt commented on cert_manager/datadog_checks/cert_manager/data/conf.yaml.example:
```diff
@@ -0,0 +1,38 @@
+
+instances:
+
+    ## @param prometheus_url - string - required
```
This file does not look up to date. I'm wondering why validation is not failing for it

sarah-witt replied:
ah it's because it is not being recognized by ddev
```
cert_manager:
    No spec found
```

sarah-witt replied:
@davidor Can you please add this code block under "assets" in the manifest.json?

```
 "configuration": {
      "spec": "assets/configuration/spec.yaml"
    },
```

davidor commented:
@sarah-witt Thanks for the review.  I've addressed your comments:
- Deleted `requirements.in`.
- Deleted `MANIFEST.in`.
- Updated `setup.py` with some things that were missing from the original one. I wasn't sure about what's no longer needed.
- Removed the `pip install -r requirements.in` line from `tox.ini`.
- Updated the `datadog-checks-base` dependency requirement to `>=25.1.0`. I don't know which version we should use exactly, so I chose this one which is the one that most integrations are using.
- Added the spec in manifest.json.
- Regenerated the example spec. My understading is that this integration should use `openmetrics_legacy` because I've seen that it accepts `prometheus_url` as a config param.

I've added each thing in a different commit.

hithwen commented:
> Reference

Has this been addressed? I think we should still bump the version to generate a new release from core, but we should definetly preserve prior history

davidor commented:
@hithwen I added a new commit that brings the changelog from integrations-extras and continues with the same version number.

FlorentClarret reviewed (approved):
Looks good to me

urseberry commented on cert_manager/README.md:
```diff
@@ -0,0 +1,54 @@
+
+### Installation
+
+The cert_manager check is included in the [Datadog Agent][3] package.
```
Shouldn't "Datadog Agent" go to a different link? [3] goes to Kubernetes Integrations Autodiscovery.

davidor replied:
Thanks. I fixed it.

FlorentClarret reviewed (approved):
👍

FlorentClarret commented on cert_manager/manifest.json:
```diff
@@ -0,0 +1,39 @@
+  "metric_to_check": "cert_manager.prometheus.health",
+  "creates_events": false,
+  "short_description": "Track all your cert-manager metrics with Datadog",
+  "guid": "17888d72-430b-4c0b-b557-e08fd96d4f1f",
```
As discussed, could you switched this id to the one we have in -extras? Thanks

davidor replied:
Done :+1:

davidor replied:
In manifest v2 there isn't a `guid` field anymore. I'm not sure if there's another field that we need to sync with extras too.

ofek commented on cert_manager/datadog_checks/cert_manager/cert_manager.py:
```diff
@@ -0,0 +1,37 @@
+# (C) Datadog, Inc. 2019-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from datadog_checks.base import OpenMetricsBaseCheck
```
I view the migration already as a massive change/starting fresh so can you please use `OpenMetricsBaseCheckV2`?

davidor replied:
I added a new commit with this

davidor commented:
The CI is failing because there's a check that detects that there's already an integration with the same name, which is expected because we know that it exists in the integrations-extras repo.

Apart from that, I pushed 3 new commits that were added after the last approval:
- Migrate to OpenMetrics V2 (as suggested by @ofek ).
- Adapt to changes in tooling.
- Migrate manifest to v2.0.0 (it's now required to make the tests pass).

ofek commented on cert_manager/datadog_checks/cert_manager/cert_manager.py:
```diff
@@ -0,0 +1,35 @@
+
+from datadog_checks.base import OpenMetricsBaseCheckV2
+
+from .metrics import ACME_METRICS, CERT_METRICS, CONTROLLER_METRICS, TYPE_OVERRIDES
```
```suggestion
from .metrics import ACME_METRICS, CERT_METRICS, CONTROLLER_METRICS
```

ofek commented on cert_manager/datadog_checks/cert_manager/cert_manager.py:
```diff
@@ -0,0 +1,35 @@
+        metric_map.update(ACME_METRICS)
+        metric_map.update(CERT_METRICS)
+
+        return {'metrics': construct_metrics_config(metric_map, TYPE_OVERRIDES)}
+
+
+def construct_metrics_config(metric_map, type_overrides):
+    metrics = []
+    for raw_metric_name, metric_name in metric_map.items():
+        config = {raw_metric_name: {'name': metric_name}}
+        if raw_metric_name in type_overrides:
+            config[raw_metric_name]['type'] = type_overrides[raw_metric_name]
+
+        metrics.append(config)
+
+    return metrics
```
```suggestion
        return {'metrics': metric_map}
```

davidor replied:
I think it should be `return {'metrics': [metric_map]}`:

```TypeError: Setting `metrics` must be an array```

ofek commented on cert_manager/datadog_checks/cert_manager/metrics.py:
```diff
@@ -0,0 +1,22 @@
+}
+
+CONTROLLER_METRICS = {
+    'certmanager_clock_time_seconds': 'clock_time',
```
```suggestion
    'certmanager_clock_time_seconds': {'name': 'clock_time', 'type': 'gauge'},
```

ofek commented on cert_manager/datadog_checks/cert_manager/metrics.py:
```diff
@@ -0,0 +1,22 @@
+    'certmanager_http_acme_client_request_count': 'http_acme_client.request',
+    'certmanager_http_acme_client_request_duration_seconds': 'http_acme_client.request.duration',
+}
+
+TYPE_OVERRIDES = {
+    'certmanager_clock_time_seconds': 'gauge',
+}
```
```suggestion
```

ofek commented on cert_manager/manifest.json:
```diff
@@ -0,0 +1,52 @@
+{
+  "manifest_version": "2.0.0",
+  "app_uuid": "a3b15f7a-5edb-4d8b-8e74-84cabf38b7e7",
```
https://github.com/DataDog/integrations-extras/blob/185c1b6cf1d896579e7a1abe695ebd75b0ef98c8/cert_manager/manifest.json#L3

```suggestion
  "app_uuid": "d8bac6db-8cf7-49ca-a4b8-643714fbc7b9",
```
